PowerEditor/src/ScintillaComponent/DocTabView.cpp

```diff
@@ -195,7 +195,7 @@ void DocTabView::bufferUpdated(Buffer * buffer, int mask)
 
 void DocTabView::setBuffer(size_t index, BufferID id)
 {
-	if (index < 0 || index >= _nbItem)
+	if (index >= _nbItem)

```

Comment:
@ozone10,

I've been wondering why index/pos is `size_t` across the project.
1. `int` should be sufficient for the number of opened tabs.
2.  `if (index < 0)` is a good "safety check".

What do you think?

Thank you.

Reply:
@Yaron10 
1. `int` is large enough, but replacing all `size_t` does not give significant performance improvement since I don't expect user to have 10000+ tabs.
2. it's good check but only for signed type.

Reply:
@ozone10,

I may not have been clear enough.

> it's good check but only for signed type.

That was my point.
This "good safety check" is removed here because _nbItem/index are `size_t`.

As you and @donho understand.

Thank you again.

Reply:
@Yaron10 
> I've been wondering why index/pos is size_t across the project.

It was `void DocTabView::setBuffer(int index, BufferID id)` hence `if (index < 0)`.
The signature of function has been changed, but `if (index < 0)` not being removed.
The changing is just for the semantic reason - the index should be size_t - an unsigned type which should never negative.

Reply:
> The changing is just for the semantic reason - the index should be size_t - an unsigned type which should never negative.

You mean to say it <del>should</del> *can* never be negative:

~~~cpp
#include <cstdio>
#include <cinttypes>
int main(void)
{
  printf("%" PRIu64, (size_t)-1);
}
~~~

Prints: `18446744073709551615`

To better explain @Yaron10's concern, a negative index won't be negative anymore; it will overflow to `ULLONG_MAX`, testing `true` when it should be `false`.

Remember how the Pascal function parser [got into infinite loops] when `targetStart` was negative? [The fix] involved changing `size_t` to a (signed) `intptr_t`.

[got into infinite loops]: https://github.com/notepad-plus-plus/notepad-plus-plus/issues/12822#issuecomment-1397862706
[The fix]: https://github.com/notepad-plus-plus/notepad-plus-plus/commit/63d1bbae78d01c9e2142e8617c78b4e93d91938d

Reply:
@donho,

Thank you for replying.

I'll add some words to @rdipardo's eloquent explanation (thank you Robert!). 

I suppose that when `if (index < 0)` was added, the main concern was not explicitly assigning `-1` to `index`.
It was the thought of (e.g.) decrementing `index` in a loop, and accidentally ending up with a negative value.

It's true that `size_t index` would not have a negative value in that case.
But getting a wrong value is still possible.

If we get that wrong value by accidentally decrementing `index` to `-1`, we can assume that `if (index >= _nbItem)` would be `true`. - But the range of possible wrong values is wider.
So, isn't the use of an unsigned variable a potential cause of trouble? Wouldn't it be better to continue using those "safety checks" with a signed variable?

Reply:
@Yaron10  @rdipardo 
Thank you for your explanation.
I see your point.
I may pay attention to this issue in the future.
For the current PR, let's make as less change as possible.

Reply:
> For the current PR, let's make as less change as possible.

*Et voilà!*

~~~diff
diff --git a/PowerEditor/gcc/makefile b/PowerEditor/gcc/makefile
index f060e860..68a2ee2e 100644
--- a/PowerEditor/gcc/makefile
+++ b/PowerEditor/gcc/makefile
@@ -37,7 +37,7 @@ LEXILLA_DIRECTORY := ../../lexilla
 LEXILLA_TARGET := liblexilla.a
 
 CXX := $(CROSS_COMPILE)g++
-CXXFLAGS := -include $(GCC_DIRECTORY)/gcc-fixes.h -std=c++20
+CXXFLAGS := -include $(GCC_DIRECTORY)/gcc-fixes.h -std=c++20 -Wno-type-limits
 RC := $(CROSS_COMPILE)windres
 RCFLAGS :=
 CPP_PATH := $(SCINTILLA_DIRECTORY)/include $(LEXILLA_DIRECTORY)/include

~~~

Seriously, though &#x2014; is a harmless GCC diagnostic really worth a permanent, wide-ranging change to the code base?

I've [given this rant before](https://github.com/notepad-plus-plus/notepad-plus-plus/issues/11728#issuecomment-114008078), but the facts remain that GCC is *not* a static analyzer, like [the one that comes free with Visual Studio](https://docs.microsoft.com/en-us/cpp/build/reference/analyze-code-analysis); it's not even a legitimate compile target, since N++ has no dependency on GTK or any other Linux-only libraries.

Reply:
@Yaron10 
I can also say using signed `int` can cause issue, because we could forget `if (index < 0)`.
Count of wrong values is same, but for unsigned variables the range is continuous and so it needs to be checked only once.

btw 
MSVC has [C4296](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-4-c4296?view=msvc-170)
but it's not enabled by default